## The simulate_drift.py script performes computer simulations using a simple Wright-Fisher model of neutral evolution. The Wright-Fisher model assumes an infinite haploid population with a constant number N of surviving offspring with only two alleles A and a at frequencies p(A) and p(a). At each generation a random number j is drawn from a binomial distribution with parameters N and p(A), resulting in a new p(A)t+1 by dividing j by N. This process is repeated for each following generation. Forward simulations were designed to match the experimental data as closely as possible. Thus, at the beginning of the simulations, the script uses the allele frequencies obtained from the base-population (defined with --base-index) to estimate p(A) and the effective population size computed for the real data to estimate the effective population size N. Our simulations were based on the same number of replicates and generations (defined with --derived-timestamps) as in the real data. In addition, we accounted for heterogeneity in coverage in the Pool-Seq data and sampled for every SNP the same number of reads as in the real data sets by drawing from a binomial distribution with the simulated allele frequency.  Additionally, one has to define minimum count, minimum and maximum coverage (note that only one absolute coverage value is possible in this script). Additionally, the effective population size needs to be defined with --populations-size. See the help within the script for further information.

## For example if you have a sync file with three replicates for the Base population, gerneration 15 and generation 30 and you want to simulate drift based on these data for the same number of generations, the command should look like this:

python /popgentools/users/martin/simulating_drift/simulate-drift.py --input input.sync --output simulated.sync --min-count 2 --min-coverage 20 --maxcoverage 500 --base-index 1,2,3 --derived-index 4,5,6,7,8,9 --derived-timestamps 15,15,15,30,30,30 --populations-size 200